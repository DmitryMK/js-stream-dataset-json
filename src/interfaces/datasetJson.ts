// Variable Types
export type ItemType =
    | 'string'
    | 'integer'
    | 'float'
    | 'double'
    | 'decimal'
    | 'boolean'
    | 'date'
    | 'time'
    | 'datetime'
    | 'URI'
    ;

// Target Variable Types
export type ItemTargetType =
    | 'integer'
    | 'decimal';

// Type of returned Object from getData
export type ItemDataObject = { [name: string]: string | number | boolean | null };

// The first item in the data array needs to be a number (itemGroupDataSeq)
export type ItemDataArray = Array<string | number | boolean | null >;

// Source System interface
export interface SourceSystem {
    // Name of the Source System
    name: string;
    // Version of the Source System
    version: string;
}


// Definition for Variable in the Dataset
export interface ItemDescription {
    // Unique identifier for Variable. Must correspond to ItemDef/@OID in Define-XML.
    itemOID: string;
    // Name for Variable
    name: string;
    // Label for Variable
    label: string;
    // Data type for Variable
    dataType: ItemType;
    // Indicates the data type into which the receiving system must transform the associated Dataset-JSON variable.
    targetDataType?: ItemType;
    // Length for Variable
    length?: number;
    // Display format supports data visualization of numeric float and date values.
    displayFormat?: string;
    // Indicates that this item is a key variable in the dataset structure. It also provides an ordering for the keys.
    keySequence?: number;
}


// Definition for Dataset-JSON
export interface Dataset {
    // Time of creation of the file containing the document.
    datasetJSONCreationDateTime: string;
    // Version of Dataset-JSON standard.
    datasetJSONVersion: string;
    // The total number of records in a dataset
    records: number;
    // The human readable name of the dataset
    name: string;
    // A short description of the dataset
    label: string;
    // Variable metadata
    columns: Array<ItemDescription>;
    // Data
    rows: Array<ItemDataArray>;
    // The date/time source database was last modified.
    dbLastModifiedDateTime? : string;
    // A unique identifier for this file.
    fileOID?: string;
    // The organization that generated the Dataset-JSON file.
    originator?: string;
    // The computer system or database management system that is the source of the information in this file.
    sourceSystem?: SourceSystem;
    // Unique identifier for Study. See ODM definition for study OID (ODM/Study/@OID).
    studyOID?: string;
    // Metadata for the data contained in the file.
    metaDataVersionOID?: string;
    // URL for a metadata file the describing the data.
    metaDataRef?: string;
    // Foreign key to ItemGroupDef.OID in Define / MDR
    itemGroupOID?: string;
}

// Metadata for the dataset
export type DatasetMetadata = Omit<Dataset, 'rows'>;
// Metadata Attributes
export type MetadataAttributes = keyof DatasetMetadata;
// Interface for checking which attributes are parsed
export type ParsedAttributes = {
    [name in MetadataAttributes]: boolean;
}
// Type of the object returned
export type DataType = 'array' | 'object';
export interface UniqueValues {
    [name: string]: (string | number | boolean | null)[];
}
