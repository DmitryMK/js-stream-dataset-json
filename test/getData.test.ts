import DatasetJson from "../src/index";
import { ItemDataArray } from "../src/interfaces/datasetJson";

test("Get 100 rows of dataset", async () => {
  const filePath = "test/data/adsl.json";

  const data = new DatasetJson(filePath);
  await data.init();
  let result: ItemDataArray[] = [];
  while(data.allRowsRead !== true) {
    let rows = await data.getData({start: data.currentPosition, length: 100});
    result = result.concat(rows as ItemDataArray[]);

  }
  expect(result).toMatchSnapshot();
});


test("Get 5 rows of dataset", async () => {
  const filePath = "test/data/adsl.json";

  const data = new DatasetJson(filePath);
  await data.init();
  const rows = await data.getData({start: 20, length: 5});
  expect(rows).toMatchSnapshot();
});

test("Get 5 rows of dataset as an object", async () => {
  const filePath = "test/data/adsl.json";

  const data = new DatasetJson(filePath);
  await data.init();
  const rows = await data.getData({start: 20, length: 5, type: "object"});
  expect(rows).toMatchSnapshot();
});

test("Get 5 rows of dataset as an object and keep only specific variables", async () => {
  const filePath = "test/data/adsl.json";

  const data = new DatasetJson(filePath);
  await data.init();
  const rows = await data.getData({start: 20, length: 5, type: "object", filterColumns: ["USUBJID", "TRTEDT", "STUDYID"]});
  expect(rows).toMatchSnapshot();
});
