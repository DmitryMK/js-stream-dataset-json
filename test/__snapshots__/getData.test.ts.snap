// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get 5 rows of NDJSON dataset 1`] = `
[
  [
    "CDISCPILOT01",
    "01-701-1211",
    "1211",
    "701",
    "701",
    "Xanomeline Low Dose",
    "Xanomeline Low Dose",
    54,
    "Xanomeline Low Dose",
    54,
    "2012-11-15",
    "2013-01-12",
    59,
    54,
    3186,
    76,
    "65-80",
    2,
    "YEARS",
    "WHITE",
    1,
    "F",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "N",
    "N",
    "Y",
    "",
    "Y",
    17.7,
    "<25",
    160,
    45.4,
    12,
    "2010-03-19",
    31.4,
    ">=12",
    "2012-10-30",
    "2012-11-15",
    "2013-01-14",
    9,
    "2013-01-14",
    "DEATH",
    "DISCONTINUED",
    "Death",
    12,
  ],
  [
    "CDISCPILOT01",
    "01-701-1234",
    "1234",
    "701",
    "701",
    "Placebo",
    "Placebo",
    0,
    "Placebo",
    0,
    "2013-03-30",
    "2013-09-22",
    177,
    0,
    0,
    69,
    "65-80",
    2,
    "YEARS",
    "WHITE",
    1,
    "M",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    23.4,
    "<25",
    181.6,
    77.1,
    18,
    "2006-08-11",
    79.3,
    ">=12",
    "2013-03-20",
    "2013-03-30",
    "2013-09-22",
    12,
    "2013-09-22",
    "COMPLETED",
    "COMPLETED",
    "",
    20,
  ],
  [
    "CDISCPILOT01",
    "01-701-1239",
    "1239",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-01-11",
    "2014-07-10",
    181,
    76.8,
    13905,
    56,
    "<65",
    1,
    "YEARS",
    "WHITE",
    1,
    "M",
    "HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    33.6,
    ">=30",
    168.9,
    95.9,
    11,
    "2003-09-23",
    123.2,
    ">=12",
    "2013-12-28",
    "2014-01-11",
    "2014-07-11",
    12,
    "2014-07-11",
    "COMPLETED",
    "COMPLETED",
    "",
    23,
  ],
  [
    "CDISCPILOT01",
    "01-701-1275",
    "1275",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-02-07",
    "2014-05-31",
    114,
    77.2,
    8802,
    61,
    "<65",
    1,
    "YEARS",
    "AMERICAN INDIAN OR ALASKA NATIVE",
    6,
    "M",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "N",
    "Y",
    "",
    "",
    24.6,
    "<25",
    167.6,
    69,
    12,
    "2011-09-17",
    28.3,
    ">=12",
    "2014-01-25",
    "2014-02-07",
    "2014-06-14",
    10,
    "2014-06-14",
    "WITHDRAWAL BY SUBJECT",
    "DISCONTINUED",
    "Withdrew Consent",
    22,
  ],
  [
    "CDISCPILOT01",
    "01-701-1287",
    "1287",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-01-25",
    "2014-07-26",
    183,
    76.3,
    13959,
    56,
    "<65",
    1,
    "YEARS",
    "WHITE",
    1,
    "F",
    "HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    23.5,
    "<25",
    154.9,
    56.3,
    14,
    "2009-10-30",
    50.6,
    ">=12",
    "2014-01-17",
    "2014-01-25",
    "2014-07-26",
    12,
    "2014-07-26",
    "COMPLETED",
    "COMPLETED",
    "",
    22,
  ],
]
`;

exports[`Get 5 rows of NDJSON dataset as an object 1`] = `
[
  {
    "AGE": 76,
    "AGEGR1": "65-80",
    "AGEGR1N": 2,
    "AGEU": "YEARS",
    "ARM": "Xanomeline Low Dose",
    "AVGDD": 54,
    "BMIBL": 17.7,
    "BMIBLGR1": "<25",
    "COMP16FL": "N",
    "COMP24FL": "N",
    "COMP8FL": "Y",
    "CUMDOSE": 3186,
    "DCDECOD": "DEATH",
    "DCSREAS": "Death",
    "DISCONFL": "Y",
    "DISONSDT": "2010-03-19",
    "DSRAEFL": "",
    "DTHFL": "Y",
    "DURDIS": 31.4,
    "DURDSGR1": ">=12",
    "EDUCLVL": 12,
    "EFFFL": "Y",
    "EOSSTT": "DISCONTINUED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 160,
    "ITTFL": "Y",
    "MMSETOT": 12,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2013-01-14",
    "RFENDTC": "2013-01-14",
    "RFSTDTC": "2012-11-15",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1211",
    "TRT01A": "Xanomeline Low Dose",
    "TRT01AN": 54,
    "TRT01P": "Xanomeline Low Dose",
    "TRT01PN": 54,
    "TRTDURD": 59,
    "TRTEDT": "2013-01-12",
    "TRTSDT": "2012-11-15",
    "USUBJID": "01-701-1211",
    "VISIT1DT": "2012-10-30",
    "VISNUMEN": 9,
    "WEIGHTBL": 45.4,
  },
  {
    "AGE": 69,
    "AGEGR1": "65-80",
    "AGEGR1N": 2,
    "AGEU": "YEARS",
    "ARM": "Placebo",
    "AVGDD": 0,
    "BMIBL": 23.4,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 0,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2006-08-11",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 79.3,
    "DURDSGR1": ">=12",
    "EDUCLVL": 18,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 181.6,
    "ITTFL": "Y",
    "MMSETOT": 20,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2013-09-22",
    "RFENDTC": "2013-09-22",
    "RFSTDTC": "2013-03-30",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1234",
    "TRT01A": "Placebo",
    "TRT01AN": 0,
    "TRT01P": "Placebo",
    "TRT01PN": 0,
    "TRTDURD": 177,
    "TRTEDT": "2013-09-22",
    "TRTSDT": "2013-03-30",
    "USUBJID": "01-701-1234",
    "VISIT1DT": "2013-03-20",
    "VISNUMEN": 12,
    "WEIGHTBL": 77.1,
  },
  {
    "AGE": 56,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 76.8,
    "BMIBL": 33.6,
    "BMIBLGR1": ">=30",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 13905,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2003-09-23",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 123.2,
    "DURDSGR1": ">=12",
    "EDUCLVL": 11,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "HISPANIC OR LATINO",
    "HEIGHTBL": 168.9,
    "ITTFL": "Y",
    "MMSETOT": 23,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2014-07-11",
    "RFENDTC": "2014-07-11",
    "RFSTDTC": "2014-01-11",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1239",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 181,
    "TRTEDT": "2014-07-10",
    "TRTSDT": "2014-01-11",
    "USUBJID": "01-701-1239",
    "VISIT1DT": "2013-12-28",
    "VISNUMEN": 12,
    "WEIGHTBL": 95.9,
  },
  {
    "AGE": 61,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 77.2,
    "BMIBL": 24.6,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "N",
    "COMP8FL": "Y",
    "CUMDOSE": 8802,
    "DCDECOD": "WITHDRAWAL BY SUBJECT",
    "DCSREAS": "Withdrew Consent",
    "DISCONFL": "Y",
    "DISONSDT": "2011-09-17",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 28.3,
    "DURDSGR1": ">=12",
    "EDUCLVL": 12,
    "EFFFL": "Y",
    "EOSSTT": "DISCONTINUED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 167.6,
    "ITTFL": "Y",
    "MMSETOT": 22,
    "RACE": "AMERICAN INDIAN OR ALASKA NATIVE",
    "RACEN": 6,
    "RFENDT": "2014-06-14",
    "RFENDTC": "2014-06-14",
    "RFSTDTC": "2014-02-07",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1275",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 114,
    "TRTEDT": "2014-05-31",
    "TRTSDT": "2014-02-07",
    "USUBJID": "01-701-1275",
    "VISIT1DT": "2014-01-25",
    "VISNUMEN": 10,
    "WEIGHTBL": 69,
  },
  {
    "AGE": 56,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 76.3,
    "BMIBL": 23.5,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 13959,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2009-10-30",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 50.6,
    "DURDSGR1": ">=12",
    "EDUCLVL": 14,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "HISPANIC OR LATINO",
    "HEIGHTBL": 154.9,
    "ITTFL": "Y",
    "MMSETOT": 22,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2014-07-26",
    "RFENDTC": "2014-07-26",
    "RFSTDTC": "2014-01-25",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1287",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 183,
    "TRTEDT": "2014-07-26",
    "TRTSDT": "2014-01-25",
    "USUBJID": "01-701-1287",
    "VISIT1DT": "2014-01-17",
    "VISNUMEN": 12,
    "WEIGHTBL": 56.3,
  },
]
`;

exports[`Get 5 rows of NDJSON dataset as an object and keep only specific variables 1`] = `
[
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2013-01-12",
    "USUBJID": "01-701-1211",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2013-09-22",
    "USUBJID": "01-701-1234",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-07-10",
    "USUBJID": "01-701-1239",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-05-31",
    "USUBJID": "01-701-1275",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-07-26",
    "USUBJID": "01-701-1287",
  },
]
`;

exports[`Get 5 rows of dataset 1`] = `
[
  [
    "CDISCPILOT01",
    "01-701-1211",
    "1211",
    "701",
    "701",
    "Xanomeline Low Dose",
    "Xanomeline Low Dose",
    54,
    "Xanomeline Low Dose",
    54,
    "2012-11-15",
    "2013-01-12",
    59,
    54,
    3186,
    76,
    "65-80",
    2,
    "YEARS",
    "WHITE",
    1,
    "F",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "N",
    "N",
    "Y",
    "",
    "Y",
    17.7,
    "<25",
    160,
    45.4,
    12,
    "2010-03-19",
    31.4,
    ">=12",
    "2012-10-30",
    "2012-11-15",
    "2013-01-14",
    9,
    "2013-01-14",
    "DEATH",
    "DISCONTINUED",
    "Death",
    12,
  ],
  [
    "CDISCPILOT01",
    "01-701-1234",
    "1234",
    "701",
    "701",
    "Placebo",
    "Placebo",
    0,
    "Placebo",
    0,
    "2013-03-30",
    "2013-09-22",
    177,
    0,
    0,
    69,
    "65-80",
    2,
    "YEARS",
    "WHITE",
    1,
    "M",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    23.4,
    "<25",
    181.6,
    77.1,
    18,
    "2006-08-11",
    79.3,
    ">=12",
    "2013-03-20",
    "2013-03-30",
    "2013-09-22",
    12,
    "2013-09-22",
    "COMPLETED",
    "COMPLETED",
    "",
    20,
  ],
  [
    "CDISCPILOT01",
    "01-701-1239",
    "1239",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-01-11",
    "2014-07-10",
    181,
    76.8,
    13905,
    56,
    "<65",
    1,
    "YEARS",
    "WHITE",
    1,
    "M",
    "HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    33.6,
    ">=30",
    168.9,
    95.9,
    11,
    "2003-09-23",
    123.2,
    ">=12",
    "2013-12-28",
    "2014-01-11",
    "2014-07-11",
    12,
    "2014-07-11",
    "COMPLETED",
    "COMPLETED",
    "",
    23,
  ],
  [
    "CDISCPILOT01",
    "01-701-1275",
    "1275",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-02-07",
    "2014-05-31",
    114,
    77.2,
    8802,
    61,
    "<65",
    1,
    "YEARS",
    "AMERICAN INDIAN OR ALASKA NATIVE",
    6,
    "M",
    "NOT HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "N",
    "Y",
    "",
    "",
    24.6,
    "<25",
    167.6,
    69,
    12,
    "2011-09-17",
    28.3,
    ">=12",
    "2014-01-25",
    "2014-02-07",
    "2014-06-14",
    10,
    "2014-06-14",
    "WITHDRAWAL BY SUBJECT",
    "DISCONTINUED",
    "Withdrew Consent",
    22,
  ],
  [
    "CDISCPILOT01",
    "01-701-1287",
    "1287",
    "701",
    "701",
    "Xanomeline High Dose",
    "Xanomeline High Dose",
    81,
    "Xanomeline High Dose",
    81,
    "2014-01-25",
    "2014-07-26",
    183,
    76.3,
    13959,
    56,
    "<65",
    1,
    "YEARS",
    "WHITE",
    1,
    "F",
    "HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    23.5,
    "<25",
    154.9,
    56.3,
    14,
    "2009-10-30",
    50.6,
    ">=12",
    "2014-01-17",
    "2014-01-25",
    "2014-07-26",
    12,
    "2014-07-26",
    "COMPLETED",
    "COMPLETED",
    "",
    22,
  ],
]
`;

exports[`Get 5 rows of dataset as an object 1`] = `
[
  {
    "AGE": 76,
    "AGEGR1": "65-80",
    "AGEGR1N": 2,
    "AGEU": "YEARS",
    "ARM": "Xanomeline Low Dose",
    "AVGDD": 54,
    "BMIBL": 17.7,
    "BMIBLGR1": "<25",
    "COMP16FL": "N",
    "COMP24FL": "N",
    "COMP8FL": "Y",
    "CUMDOSE": 3186,
    "DCDECOD": "DEATH",
    "DCSREAS": "Death",
    "DISCONFL": "Y",
    "DISONSDT": "2010-03-19",
    "DSRAEFL": "",
    "DTHFL": "Y",
    "DURDIS": 31.4,
    "DURDSGR1": ">=12",
    "EDUCLVL": 12,
    "EFFFL": "Y",
    "EOSSTT": "DISCONTINUED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 160,
    "ITTFL": "Y",
    "MMSETOT": 12,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2013-01-14",
    "RFENDTC": "2013-01-14",
    "RFSTDTC": "2012-11-15",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1211",
    "TRT01A": "Xanomeline Low Dose",
    "TRT01AN": 54,
    "TRT01P": "Xanomeline Low Dose",
    "TRT01PN": 54,
    "TRTDURD": 59,
    "TRTEDT": "2013-01-12",
    "TRTSDT": "2012-11-15",
    "USUBJID": "01-701-1211",
    "VISIT1DT": "2012-10-30",
    "VISNUMEN": 9,
    "WEIGHTBL": 45.4,
  },
  {
    "AGE": 69,
    "AGEGR1": "65-80",
    "AGEGR1N": 2,
    "AGEU": "YEARS",
    "ARM": "Placebo",
    "AVGDD": 0,
    "BMIBL": 23.4,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 0,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2006-08-11",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 79.3,
    "DURDSGR1": ">=12",
    "EDUCLVL": 18,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 181.6,
    "ITTFL": "Y",
    "MMSETOT": 20,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2013-09-22",
    "RFENDTC": "2013-09-22",
    "RFSTDTC": "2013-03-30",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1234",
    "TRT01A": "Placebo",
    "TRT01AN": 0,
    "TRT01P": "Placebo",
    "TRT01PN": 0,
    "TRTDURD": 177,
    "TRTEDT": "2013-09-22",
    "TRTSDT": "2013-03-30",
    "USUBJID": "01-701-1234",
    "VISIT1DT": "2013-03-20",
    "VISNUMEN": 12,
    "WEIGHTBL": 77.1,
  },
  {
    "AGE": 56,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 76.8,
    "BMIBL": 33.6,
    "BMIBLGR1": ">=30",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 13905,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2003-09-23",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 123.2,
    "DURDSGR1": ">=12",
    "EDUCLVL": 11,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "HISPANIC OR LATINO",
    "HEIGHTBL": 168.9,
    "ITTFL": "Y",
    "MMSETOT": 23,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2014-07-11",
    "RFENDTC": "2014-07-11",
    "RFSTDTC": "2014-01-11",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1239",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 181,
    "TRTEDT": "2014-07-10",
    "TRTSDT": "2014-01-11",
    "USUBJID": "01-701-1239",
    "VISIT1DT": "2013-12-28",
    "VISNUMEN": 12,
    "WEIGHTBL": 95.9,
  },
  {
    "AGE": 61,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 77.2,
    "BMIBL": 24.6,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "N",
    "COMP8FL": "Y",
    "CUMDOSE": 8802,
    "DCDECOD": "WITHDRAWAL BY SUBJECT",
    "DCSREAS": "Withdrew Consent",
    "DISCONFL": "Y",
    "DISONSDT": "2011-09-17",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 28.3,
    "DURDSGR1": ">=12",
    "EDUCLVL": 12,
    "EFFFL": "Y",
    "EOSSTT": "DISCONTINUED",
    "ETHNIC": "NOT HISPANIC OR LATINO",
    "HEIGHTBL": 167.6,
    "ITTFL": "Y",
    "MMSETOT": 22,
    "RACE": "AMERICAN INDIAN OR ALASKA NATIVE",
    "RACEN": 6,
    "RFENDT": "2014-06-14",
    "RFENDTC": "2014-06-14",
    "RFSTDTC": "2014-02-07",
    "SAFFL": "Y",
    "SEX": "M",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1275",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 114,
    "TRTEDT": "2014-05-31",
    "TRTSDT": "2014-02-07",
    "USUBJID": "01-701-1275",
    "VISIT1DT": "2014-01-25",
    "VISNUMEN": 10,
    "WEIGHTBL": 69,
  },
  {
    "AGE": 56,
    "AGEGR1": "<65",
    "AGEGR1N": 1,
    "AGEU": "YEARS",
    "ARM": "Xanomeline High Dose",
    "AVGDD": 76.3,
    "BMIBL": 23.5,
    "BMIBLGR1": "<25",
    "COMP16FL": "Y",
    "COMP24FL": "Y",
    "COMP8FL": "Y",
    "CUMDOSE": 13959,
    "DCDECOD": "COMPLETED",
    "DCSREAS": "",
    "DISCONFL": "",
    "DISONSDT": "2009-10-30",
    "DSRAEFL": "",
    "DTHFL": "",
    "DURDIS": 50.6,
    "DURDSGR1": ">=12",
    "EDUCLVL": 14,
    "EFFFL": "Y",
    "EOSSTT": "COMPLETED",
    "ETHNIC": "HISPANIC OR LATINO",
    "HEIGHTBL": 154.9,
    "ITTFL": "Y",
    "MMSETOT": 22,
    "RACE": "WHITE",
    "RACEN": 1,
    "RFENDT": "2014-07-26",
    "RFENDTC": "2014-07-26",
    "RFSTDTC": "2014-01-25",
    "SAFFL": "Y",
    "SEX": "F",
    "SITEGR1": "701",
    "SITEID": "701",
    "STUDYID": "CDISCPILOT01",
    "SUBJID": "1287",
    "TRT01A": "Xanomeline High Dose",
    "TRT01AN": 81,
    "TRT01P": "Xanomeline High Dose",
    "TRT01PN": 81,
    "TRTDURD": 183,
    "TRTEDT": "2014-07-26",
    "TRTSDT": "2014-01-25",
    "USUBJID": "01-701-1287",
    "VISIT1DT": "2014-01-17",
    "VISNUMEN": 12,
    "WEIGHTBL": 56.3,
  },
]
`;

exports[`Get 5 rows of dataset as an object and keep only specific variables 1`] = `
[
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2013-01-12",
    "USUBJID": "01-701-1211",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2013-09-22",
    "USUBJID": "01-701-1234",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-07-10",
    "USUBJID": "01-701-1239",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-05-31",
    "USUBJID": "01-701-1275",
  },
  {
    "STUDYID": "CDISCPILOT01",
    "TRTEDT": "2014-07-26",
    "USUBJID": "01-701-1287",
  },
]
`;

exports[`Get one row of NDJSON dataset 1`] = `
[
  [
    "CDISCPILOT01",
    "01-701-1015",
    "1015",
    "701",
    "701",
    "Placebo",
    "Placebo",
    0,
    "Placebo",
    0,
    "2014-01-02",
    "2014-07-02",
    182,
    0,
    0,
    63,
    "<65",
    1,
    "YEARS",
    "WHITE",
    1,
    "F",
    "HISPANIC OR LATINO",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "Y",
    "",
    "",
    "",
    25.1,
    "25-<30",
    147.3,
    54.4,
    16,
    "2010-04-30",
    43.9,
    ">=12",
    "2013-12-26",
    "2014-01-02",
    "2014-07-02",
    12,
    "2014-07-02",
    "COMPLETED",
    "COMPLETED",
    "",
    23,
  ],
]
`;
